toolbox/bulk_processing/build_address_update_from_sample.py

```diff
@@ -0,0 +1,83 @@
+import argparse
+import csv
+from copy import copy
+from pathlib import Path
+
+import requests
+from requests import HTTPError
+
+from toolbox.bulk_processing.address_update_processor import AddressUpdateProcessor
+from toolbox.config import Config
+
+
+def generate_address_update_file(file_to_process: Path):
+    output_address_update_file = Path(f'{Config.BULK_ADDRESS_UPDATE_FILE_PREFIX}{file_to_process.stem}.csv')
+
+    case_id_fieldname = 'CASE_ID'
+    address_update_fieldnames = list(AddressUpdateProcessor.schema.keys())
+    expected_input_fields = set(address_update_fieldnames)
+    expected_input_fields.remove(case_id_fieldname)
+    fields_to_copy = copy(expected_input_fields)
+
+    # There are columns that are on the sample file but not address update,
+    # we expect to see them but ignore the values here
+    expected_input_fields.update({'REGION', 'ADDRESS_TYPE', 'ADDRESS_LEVEL'})
+
+    with open(file_to_process) as open_file_to_process:
+        file_reader = csv.DictReader(open_file_to_process, delimiter=',')
+        actual_fieldnames = set(file_reader.fieldnames)
+        if actual_fieldnames != expected_input_fields:
+            output_address_update_file.unlink()
+            raise ValueError(f'Invalid header, '
+                             f'missing field names: {expected_input_fields.difference(actual_fieldnames)}, '
+                             f'unexpected fieldnames: {actual_fieldnames.difference(expected_input_fields)}')
+
+        with open(output_address_update_file, 'w') as open_output_file:
+            output_writer = csv.DictWriter(open_output_file, fieldnames=address_update_fieldnames)
+            output_writer.writeheader()
+
+            for line_number, row in enumerate(file_reader, 1):
+                try:
+                    case_id = get_case_id_from_case_api(row['UPRN'], line_number)
+                except Exception:
+                    output_address_update_file.unlink()
+                    raise
+                address_update_row = {k: row[k] for k in fields_to_copy}
+                address_update_row[case_id_fieldname] = case_id
+                output_writer.writerow(address_update_row)
+
+    return output_address_update_file
+
+
+def get_case_id_from_case_api(uprn, line_number):
```

Comment:
I estimate this job could take over 2 hours to run (20 milliseconds per roundtrip times 414,000)... in which case a glitch with the Case API would be annoying. Should this retry? Kinda a nice to have I guess

Reply:
Fair point, I was also unsure about failing out on errors. On the one hand the file should have been pre-validated, on the other failing the whole thing in the 100th minute when 99.9% of the data could have been fine is frustrating.

Reply:
up to you... plenty of our other scripts have similar glitch problems